yogstation/code/modules/antagonists/darkspawn/darkspawn_antag.dm

```diff
@@ -0,0 +1,635 @@
+		return
+	var/delay = (1/psi_per_second) SECONDS
+	addtimer(CALLBACK(src, PROC_REF(regenerate_psi)), delay, TIMER_UNIQUE) //tick it up very quickly
+
+///temporarily block psi regeneration
+/datum/antagonist/darkspawn/proc/block_psi(duration = 5 SECONDS, identifier)
+	if(!identifier)
+		return
+	ADD_TRAIT(src, TRAIT_DARKSPAWN_PSIBLOCK, identifier)
+	if(owner.current)
+		owner.current.throw_alert("psiblock", /atom/movable/screen/alert/psiblock)
+	addtimer(CALLBACK(src, PROC_REF(unblock_psi), identifier), duration, TIMER_UNIQUE | TIMER_OVERRIDE)
+
+/datum/antagonist/darkspawn/proc/unblock_psi(identifier)
+	REMOVE_TRAIT(src, TRAIT_DARKSPAWN_PSIBLOCK, identifier)
+	if(!HAS_TRAIT(src, TRAIT_DARKSPAWN_PSIBLOCK) && owner.current)
+		owner.current.clear_alert("psiblock")
+
+/datum/antagonist/darkspawn/proc/update_psi_hud()
+	if(!owner.current || !owner.current.hud_used)
+		return
+	var/atom/movable/screen/counter = owner.current.hud_used.psi_counter
+	counter.maptext = ANTAG_MAPTEXT(psi, COLOR_DARKSPAWN_PSI)
+
+////////////////////////////////////////////////////////////////////////////////////
+//-----------------------------------Divulge--------------------------------------//
+////////////////////////////////////////////////////////////////////////////////////
+/datum/antagonist/darkspawn/proc/divulge(forced = FALSE)
+	if(darkspawn_state >= DARKSPAWN_DIVULGED)
+		return FALSE
+		
+	var/mob/living/carbon/human/user = owner.current
+
+	if(!user || !istype(user))//sanity check
+		return
+
+	if(!picked_class) //you didn't pick, now it gets forced on you
+		var/list/classes = list()
+		for(var/datum/component/darkspawn_class/class as anything in subtypesof(/datum/component/darkspawn_class))
+			if(initial(class.choosable))
+				classes |= class
+
+		var/chosen = pick(classes)
+
+		picked_class = owner.AddComponent(chosen)
+	
+	owner.assigned_role = picked_class.name //they stop being whatever job they were the moment they divulge
+
+	if(forced)
+		owner.current.visible_message(
+			span_boldwarning("[owner.current]'s skin sloughs off, revealing black flesh covered in symbols!"), 
+			span_userdanger("You have forcefully divulged!"))
+
+	for(var/datum/action/cooldown/spell/spells in user.actions) //remove the ability that triggers this
+		if(istype(spells, /datum/action/cooldown/spell/divulge))
+			spells.Remove(user)
+			qdel(spells)
+
+	disguise_name = user.real_name //keep track of the old name
+	user.fully_heal()
+	user.set_species(/datum/species/shadow/darkspawn)
+	ADD_TRAIT(user, TRAIT_SPECIESLOCK, "darkspawn divulge") //prevent them from swapping species which can fuck stuff up
+	user.update_appearance(UPDATE_OVERLAYS)
+
+	show_to_ghosts = TRUE
+	var/processed_message = span_velvet("<b>\[Mindlink\] [disguise_name] has removed their human disguise and is now [user.real_name].</b>")
+	for(var/T in GLOB.alive_mob_list)
+		var/mob/M = T
+		if(is_darkspawn_or_thrall(M) || (ROLE_DARKSPAWN in M.faction))
+			to_chat(M, processed_message)
```

Comment:
Same as before, use ``as anything`` but please look into things like ``get_antag_minds``

Reply:
this one specifically has to be mobs, as they can spawn a simplemob mob that doesn't get an antag datum, but is on their team